### ISSUE TITLE
Support container queries

### ISSUE THREAD

**Issue by tanhauhau** [csstree/csstree#174]
Currently using container queries leads to a parse error:

```css
@container (min-width: 400px) {
  .child {
    display: flex;
    flex-wrap: wrap;
  }
}
```

ParseError: RightParenthesis is expected

Container Queries:
- https://drafts.csswg.org/css-contain-3/
- https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Container_Queries

**Comment by mherchel** [comment]
I'm also running into this issue attempting to implement container queries with [USWDS](https://designsystem.digital.gov/)

**Comment by georgecrawford** [comment]
Is there any hint from the maintainer(s) or community about when this might be implemented, or alternatively a bit of help as to how to implement it for a newbie? I'd love to get `@container` support into Svelte!

**Comment by arxpoetica** [comment]
Container queries are starting to land in all browsers.

**Comment by v-erse** [comment]
> Container queries are starting to land in all browsers, Firefox pending.

Container queries enabled on Firefox Nightly -> https://bugzilla.mozilla.org/show_bug.cgi?id=1801123

**Comment by typhonrt** [comment]
This is definitely a hard blocking issue for a forward looking UI framework I'm working on for Svelte as I need to container queries to be scoped to Svelte components / no workarounds. I have taken a first pass at implementing container query support in `csstree` and enabling it in Svelte compiler is trivial (one line adjustment). Everything works great. 

I haven't written extensive tests yet as I'd like to work with @lahmatiy to get this addition properly sorted out as it isn't exactly a trivial addition to `csstree`. I have only written the parsing support and no lexer validation of the parsed `@container` prelude features. I modeled CQ support based on the media query support, but of course it is a little bit different. There is support for style queries and range queries in addition to standard queries. It should be noted that `csstree` also needs to be updated for range syntax in media queries as well. 

Here is my fork:
https://github.com/typhonjs-svelte/csstree

In particular these are the new files:
https://github.com/typhonjs-svelte/csstree/blob/master/lib/syntax/node/Comparison.js
https://github.com/typhonjs-svelte/csstree/blob/master/lib/syntax/node/ContainerFeature.js
https://github.com/typhonjs-svelte/csstree/blob/master/lib/syntax/node/ContainerFeatureRange.js
https://github.com/typhonjs-svelte/csstree/blob/master/lib/syntax/node/ContainerFeatureStyle.js
https://github.com/typhonjs-svelte/csstree/blob/master/lib/syntax/node/ContainerQuery.js

A new method in `create.js`: 
https://github.com/typhonjs-svelte/csstree/blob/master/lib/parser/create.js#L158-L169

All of the new files / edit points have `/* TODO: NEW CQ */` comment. 

-------------------------

TODO: Extensive test case generation. I have manually verified that things parse correctly with many variations to the different query types. There is no lexer support added to validate types of queries and mismatched nodes.  IE a range query improperly formatted: `@container (<= width >= 0)` will parse, but not verified in lexer currently. 

-------------------------

Potential other improvements. Currently I'm generating a new `Comparison` node: `'<', '<=', '=', '>=', and '<='` through parsing delimeters / Delim tokens. It might be a good idea to add a comparison type to the tokenizer to make things more efficient and share that aspect with media queries. 

-------------------------

Example AST output of simple query: 
```
@container (min-width: 0) {
    .rule {
        box-shadow: var(--something, fallback);
    }
}
```

Result: 
[stringify.ast.zip](https://github.com/csstree/csstree/files/10098992/stringify.ast.zip)

------------

Example AST output for style query w/ container name: 
```
@container container-name style(background-color: blue) {
    .rule {
        box-shadow: var(--something, fallback);
    }
}
```

Result: 
[stringify.ast.zip](https://github.com/csstree/csstree/files/10098997/stringify.ast.zip)

------------

Example AST output for range query w/ container name: 
```
@container container-name (0 >= width < 200px) {
    .rule {
        box-shadow: var(--something, fallback);
    }
}
```

Result: 
[stringify.ast.zip](https://github.com/csstree/csstree/files/10099012/stringify.ast.zip)

**Comment by typhonrt** [comment]
Thanks for getting in touch @lahmatiy. Indeed my proof of concept tried to touch / modify the least amount of code and use what I found vs a potential refactor. It definitely seems like a refactor to serve similar purposes across related at-rules makes the most sense. I know it can be quite a bit of work to keep things updated. I spent quite some time on a massive AST parsing / complexity analysis package for JS / ESM / TS that still seems to be used, but is getting long in the tooth, so to speak, for sure. I appreciate your development of `csstree`. 

Do you think in the interim that you can post a release that doesn't parse or evaluate the @container prelude. For instance if you currently look at ASTExplorer and the various CSS parsers only PostCSS doesn't crash / fail currently and it doesn't attempt to parse the @container prelude: 
https://astexplorer.net/#/gist/244e2fb4da940df52bf0f4b94277db44/e3cc567c28b85c5483bda47c4133788b51ac3412

The problem is that with the current state of things a parse error is thrown and Svelte attaches an `onParseError` handler causing things to fail even though it would be perfectly acceptable to receive the prelude as a `Raw` node. To get container queries to work in Svelte it isn't dependent on the prelude parsing or evaluating that part of the AST. It only is concerned with the block section / inner content.

Getting container queries to work in Svelte then amounts to `csstree` not causing a parse error and simply changing this one line in Svelte adding the node.name check for 'container': 
`if (this.node.name === 'container' || this.node.name === 'media' || this.node.name === 'supports' || this.node.name === 'layer')`

**Comment by lahmatiy** [comment]
Btw, you can extend `css-tree` functionality (i.e. syntax) in many ways without forking it as a library. To achieve it you need to use a `fork()` method:

```js
import { fork } from 'css-tree';

const mySyntax = fork({
    atrule: { // extend or override at-rule dictionary
        container: { // the name of an at-rule, as other default at-rules are defined, e.g. https://github.com/csstree/csstree/blob/master/lib/syntax/atrule/media.js
            parse: {
                prelude() { ... },
                body() { ... }
            }
        }
    },
    definitions: {
        atrules: { // in case you want to extend syntax matching and validation
            container: {
                prelude: '[ <container-name> ]? <container-condition>'
            }
        },
        types: { // don't forget to specify all related types if you introduce something new
            'container-name': '<custom-ident>',
            // ...
        }
    },
    node: { // extend node types
        MyNodeType: { // it will be available in the parser as this.MyNodeType()
            name: 'MyNodeType',
            structure: { /* ... */ },
            parse() { },
            generate() { }
        }
    }
});

// mySyntax has all the main syntax related methods, see https://github.com/csstree/csstree/blob/master/lib/index.js#L14-L30
const ast = mySyntax.parse('@container example {}')
const css = mySyntax.generate(ast);
```

**Comment by typhonrt** [comment]
@lahmatiy That is great to know about the adhoc extension. An approach like this could be workable in a project that I control a bit more, but in relation to getting container query support into Svelte I gather that the core Svelte maintainers are only interested in bumping the `css-tree` version and adding a simple modification to one line. 

Let me know if there is anything I can do to help w/ the `queries` branch whether that is testing out the current state, writing tests, or something else / etc.

**Comment by jogibear9988** [comment]
> Hi folks!
> 
> Sorry for a long delay for this issue. I thought a lot on the problem and made some experiments and steps forward, but not completed it yet. There are some details on the feature status.
> 
> The container queries is not a completely new feature from a parser perspective. Beside the container queries we have queries in `@media` and `@supports` at-rules as well as in `@import` prelude. All of these queries are relatively behave the same, the difference mostly in a type of condition's terms.

but you know, that container queries also could be nested?

**Comment by benmccann** [comment]
Thanks for your efforts around this @lahmatiy. Is there anything that external contributors could lend a hand with?

**Comment by benmccann** [comment]
Just FYI, Svelte has merged container query support by using a patched version of csstree for now: https://github.com/sveltejs/svelte/pull/8275. Hopefully we'll be able to unfork it in the future

**Comment by typhonrt** [comment]
This is the `fork` functionality that was mentioned in [this comment](https://github.com/csstree/csstree/issues/174#issuecomment-1337001937).

I'm definitely still down to help out with anything relating the queries branch.

**Comment by Mister-Hope** [comment]
I am not sure, but can anyone achieve this, currently this is supported by all browsers:

![image](https://github.com/csstree/csstree/assets/33315834/e5032fa2-881c-4217-b3ce-98c5bb7e8328)

While this can create issues downstream like: https://github.com/lahmatiy/postcss-csso/issues/26

**Comment by Mister-Hope** [comment]
Edited: sorry for the noise, but I just realize this is already supported in https://github.com/csstree/csstree/commit/c426179c36649e73ffda63cc3a8c22b7430ef021.

@lahmatiy Is there any release plan? As the recent commit is already 2 weeks old.

**Comment by IdosApplitools** [comment]
I am also encountering this issue when trying to parse the condition of the container rule: 
```
parse("(min-width: 400px)", {context: 'atrulePrelude', atrule: 'container'})
```

Any ETA for support?

**Comment by lahmatiy** [comment]
Fixed in [v3.0.0](https://github.com/csstree/csstree/releases/tag/v3.0.0)
